### PR TITLE
Add Terraform Engine integration tests

### LINKED ISSUE #852: Add a basic integration test to deployment team.hcl at presubmit

### PR DESCRIPTION
**Pull request by xingao267**
Fixes #852

The `integration-tests` build is configured to be only triggered when a repository collaborator or owner comments `/gcbrun`.

### CODE REVIEW

File: examples/tfengine/modules/team.hcl [hunk 1]
```diff
@@ -103,13 +103,18 @@ template "project_networks" {
...
+        display_name = "Bastion Accessor Service Account"
+      }]
       bastion_hosts = [{
         name           = "bastion-vm"
         network        = "$${module.network.network.network.self_link}"
         subnet         = "$${module.network.subnets[\"{{.default_location}}/bastion-subnet\"].self_link}"
         image_family   = "ubuntu-2004-lts"
         image_project  = "ubuntu-os-cloud"
-        members        = ["group:{{.prefix}}-bastion-accessors@{{.domain}}"]
+        members        = ["serviceAccount:$${google_service_account.bastion_accessor.email}"] # Placeholder for testing.
```
Comment (MartinPetkov) [inline examples/tfengine/modules/team.hcl:117]: How come this needs to be a serviceAccount and not a group?

File: examples/tfengine/modules/team.hcl [hunk 2]
```diff
@@ -382,13 +391,10 @@ EOF
...
           }
         }]
-        retention_policy = {
-          retention_period = 86400 # 1 day.
-        }
-        iam_members = [{
-          role   = "roles/storage.objectViewer"
-          member = "group:{{.prefix}}-data-viewers@{{.domain}}"
-        }]
+        # Skip for integration tests.
```
Comment (MartinPetkov) [inline examples/tfengine/modules/team.hcl:394]: Nit: I would rephrase slightly: "This is skipped for integration tests for <reason> [I assume so resources can be destroyed after], uncomment when writing a production template."

File: tests/tfengine/integration_test.go [hunk 3]
```diff
@@ -17,8 +17,11 @@
 //  - The environment variable RUN_INTEGRATION_TEST must be set to "true".
 //  - The environment variables BILLING_ACCOUNT and FOLDER_ID must be set.
```
Comment (MartinPetkov) [inline tests/tfengine/integration_test.go:18]: And DOMAIN must also be set

**Review by MartinPetkov — APPROVED** [review]
Overall LGTM.

Just one suggestion, if the destroy fails I think we'd need to destroy the resources manually/debug. It would be good to also add either a README for how to do this manually or a manually-triggered GCB job to do cleanup (or both), which I can imagine would be "1) Generate configs the same way as the integration tests do 2) Run importer 3) Run destroy again".

This is unlikely to happen though so I'm fine just leaving it for the future. I'll open an issue.

**Comment by MartinPetkov** [comment]
Oh, also, I'm wondering, will this integration test clash if multiple PRs are opened at the same time?

**Comment by MartinPetkov** [comment]
> Oh, also, I'm wondering, will this integration test clash if multiple PRs are opened at the same time?

I looked again and I think this line prevents this from happening:
```
prefix := fmt.Sprintf("dpt%v", time.Now().Unix())
```

File: build/presubmit_int.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  id: Terraform version
+
+- name: "${_CFT_TOOLS_CONTAINER}"
+  entrypoint: "go"
+  args: ["test", "./tests/tfengine/...", "-v", "-timeout", "60m"]
+  waitFor: ["-"]
+  id: Terraform Engine integration tests
+  env:
+  - "RUN_INTEGRATION_TEST=true"
+  - "BILLING_ACCOUNT=$_BILLING_ACCOUNT"
```
[the remark this answers is not in the document]
Comment (xingao267) [reply build/presubmit_int.yaml]: Done

File: build/presubmit_int.yaml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+# Terraform version used is 0.14.8.
+substitutions:
+  _CFT_TOOLS_CONTAINER: "gcr.io/cloud-foundation-cicd/cft/developer-tools@sha256:8b0d85766044b0e4701729f9e70ee78dcecdb8688bd86915eb2a1cb00580ba1e"
+
+timeout: 3600s
+
+options:
+  env: ["GOPATH=/go"]
+  machineType: "N1_HIGHCPU_32"
```
[the remark this answers is not in the document]
Comment (xingao267) [reply build/presubmit_int.yaml]: Let me try with `N1_HIGHCPU_8` and report back

File: examples/tfengine/modules/team.hcl (hunk 1 above)
Comment (xingao267) [reply examples/tfengine/modules/team.hcl:117]: because otherwise we need to supply an existing group in the test. Using a service account created on the fly to avoid that. It's also how CFT's tests do it for IAM members.

File: examples/tfengine/modules/team.hcl (hunk 2 above)
Comment (xingao267) [reply examples/tfengine/modules/team.hcl]: Done

File: tests/tfengine/integration_test.go (hunk 3 above)
Comment (xingao267) [reply tests/tfengine/integration_test.go]: Done

**Comment by xingao267** [comment]
> Overall LGTM.
> 
> Just one suggestion, if the destroy fails I think we'd need to destroy the resources manually/debug. It would be good to also add either a README for how to do this manually or a manually-triggered GCB job to do cleanup (or both), which I can imagine would be "1) Generate configs the same way as the integration tests do 2) Run importer 3) Run destroy again".
> 
> This is unlikely to happen though so I'm fine just leaving it for the future. I'll open an issue.

Good point. That can happen. We can simply delete the projects in the testing folder manually for now.

File: build/presubmit_int.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (xingao267) [reply build/presubmit_int.yaml]: NI_HIGHCPU_8 is enough. Almost no difference in time.

**Comment by MartinPetkov** [comment]
Just FYI the integration tests take 36 min 47 sec with just 8 CPUs. Seems like it's better to keep it with the higher resources, which seemed to take much less time even if it may take longer for the build to start.

**Comment by xingao267** [comment]
with 32, it takes about the same time.
